### ISSUE TITLE
AdaShift for Transformer?

### ISSUE THREAD

**Issue by AranKomat** [ZhimingZhou/AdaShift-LGANs-MaxGP-refactored#1]
I tried AdaShift on Transformer for large-scale language modeling, but so far it's not working well. Given the significant performance gain over Adam on NMT with LSTM, it's worth trying this direction. Have you got any result or any idea how this can be improved?

**Comment by ZhimingZhou** [comment]
Hi, Aran, 

Thanks for your interest in AdaShift. We haven't try AdaShift on Transformer, but we think it is worth trying. 

Would you provide the detailed parameter settings and results of AdaShift and some reference algorithm (e.g., Adam)? And then we will be able to analysis why AdaShift currently does not work well in your setting. 

Note that the most sensitive parameter should be the learning rate and the operation phi also has a big influence on the performance. More fine-grained tuning would involve n, beta_1 and beta_2. But if it does not work well, we would suggest to set n=1, beta_1=0, and beta_2=0.999, which is usually not a bad choice. That is, disable first moment estimation and use a large beta_2 to make v_t stable. 

Best regards,
Zhiming Zhou

**Comment by AranKomat** [comment]
Thanks for your reply. To be honest, since I've been using PyTorch implementation of Transformer, I used the [PyTorch implementation of AdaShift](https://github.com/MichaelKonobeev/adashift/blob/master/adashift/optimizers.py), which successfully reproduced your result on various tasks, including NMT with LSTM. This could be the reason why it didn't work well.

I'm using the inverse square root learning schedule, which is the standard of Transformer. Also, I tried various constant multiples of this learning rate. I tried both (beta_1, beta_2) = (0, 0.9) and (0.9, 0.999) (not (0, 0.999) yet). Notably, since the dataset I'm using is extremely large, use of dropout harms the performance in this case (so set p=0). 

For the regular Transformer on language modeling, the test loss becomes less than 7 after two hundred iterations. But for my settings with AdaShift, it floats around 7 even after thousands of iterations. 

Thanks for your advice. I'll keep trying various hyperparameters. If it doesn't work after that, then I'll conclude that maybe the PyTorch implementation contains some flaw unlike yours.

**Comment by ZhimingZhou** [comment]
I would suspect the problem lies in the learning rate schedule. 

AdaShift requires a quite different learning rate from Adam, especially when you use the reduce_max operation as phi, which according to our experiments usually requires a learning rate which is around ten times as the one used for Adam. 

Given it currently totally does not work, I would suggest you to use a constant learning rate and try different magnitudes, e.g., 1e-1, 1e-2, 1e-3, 1e-4. 

BTW. I have a brief check on the PyTorch implementation which seems to be correct. 

And, in this repo, we also provide some other choices of phi, e.g., max_mean. You may also try it. 

Best regards,
Zhiming Zhou

**Comment by AranKomat** [comment]
Thank you very much for your advice. I've tried what you suggested, but a similar problem persists. I hope someone else will be able to resolve the problem.

**Comment by ZhimingZhou** [comment]
Sorry to hear that. If you have any progress, please also let me know. 

Best regards,
Zhiming Zhou
